File: message-queue/rust/src/lib/runtime.rs
use std::error;
use std::collections::HashMap;

use chrono::prelude::*;
use uuid::Uuid;
use platform_info::{PlatformInfo, PlatformInfoAPI, UNameAPI};

use rustc_version_runtime;

use crate::Config;
use crate::prometheus::Prometheus;

#[derive(Debug)]
pub struct Runtime {
    pub ctime: DateTime<Local>,
    pub uuid: Uuid,
    pub lang: String,
    pub lang_version: rustc_version_runtime::Version,
    pub runtime: rustc_version_runtime::VersionMeta,
    pub os: String,
    pub kernel: String,
    pub arch: String,
}

impl Runtime {
    pub fn new() -> Result<Self,Box<dyn error::Error>> {
        let r = PlatformInfo::new();
        if let Err(e) = r {
            // work around PlatformInfo::new() returning
            // <dyn Error + Send + sync>
            // which makes it unusable with `?` operator since we return just error.
            // wtf?
            return Err(e);
        }
        let pi = r.unwrap();
        return Ok(Self{
            ctime: Local::now(),
            uuid: Uuid::new_v4(),
            lang: "rust".to_string(),
            lang_version: rustc_version_runtime::version(),
            runtime: rustc_version_runtime::version_meta(),
            os: pi.sysname().to_str().unwrap().to_string(),
            kernel: pi.release().to_str().unwrap().to_string(),
            arch: pi.machine().to_str().unwrap().to_string()
        });
    }

    pub fn to_map(&self) -> HashMap<String,String> {
        let rt_string = format!(
            "{}{}",
            self.runtime.short_version_string,
            match self.runtime.llvm_version.as_ref() {
                Some(x) => { format!(" llvm {}", x.to_string()) }
                None => { "".to_string() }
            }
        );

        let mut m = HashMap::new();
        m.insert("ctime".to_string(), self.ctime.to_rfc3339());
        m.insert("uuid".to_string(), self.uuid.to_string());
        m.insert("lang".to_string(), self.lang.to_owned());
        m.insert("lang_version".to_string(), self.lang_version.to_string());
        m.insert("runtime".to_string(), rt_string);
        m.insert("os".to_string(), self.os.to_owned());
        m.insert("kernel".to_string(), self.kernel.to_owned());
        m.insert("arch".to_string(), self.arch.to_owned());
        return m;
    }
}

pub struct Instance {
    pub config: Config,
    pub runtime: Runtime,
    pub prometheus: Prometheus,
}

impl Instance {
    pub fn new() -> Result<Self,Box<dyn error::Error>> {
        let r = Runtime::new()?;
        let p = Prometheus::new(&r)?;
        return Ok(
            Self{
                config: Config::new()?,
                runtime: r,
                prometheus: p,
            }
        );
    }
}

File: message-queue/golang/src/core/runtime.go
package core

import (
	"github.com/bensallen/modscan/pkg/uname"
	"github.com/google/uuid"
	"runtime"
	"time"
)

type Runtime struct {
	Ctime        time.Time
	Uuid         uuid.UUID
	Lang         string
	Lang_version string
	Runtime      string
	Os           string
	Kernel       string
	Arch         string
}

func NewRuntime() (*Runtime, error) {
	uuid, err := uuid.NewUUID()
	if err != nil {
		return nil, err
	}

	uname, err := uname.New()
	if err != nil {
		return nil, err
	}

	r := &Runtime{
		time.Now(),
		uuid,
		"golang",
		runtime.Version(),
		"golang",
		runtime.GOOS,
		uname.Release(),
		runtime.GOARCH,
	}

	return r, nil
}

func (r Runtime) Map() map[string]string {
	return map[string]string{
		"ctime":        r.Ctime.Format(time.RFC3339),
		"uuid":         r.Uuid.String(),
		"lang":         r.Lang,
		"lang_version": r.Lang_version,
		"runtime":      r.Runtime,
		"os":           r.Os,
		"kernel":       r.Kernel,
		"arch":         r.Arch,
	}
}

func RuntimeFieldNames() []string {
	return []string{"ctime", "uuid", "lang", "lang_version", "runtime", "os", "kernel", "arch"}
}
